src/js/components/color-palette.react.js

```diff
@@ -0,0 +1,209 @@
+
+        // An approximation of x based on size of widget and place within array
+        let x = (this.state.colors.length - i) * COLOR_WIDTH + HORIZONTAL_POSITION_BUFFER;
+        this.setState({ x: (screenWidth - x) });
+        this.setState({ y: (screenHeight - VERTICAL_POSITION_BUFFER) });
+
+        // Log the currentColor being edited + the position of the color within our internal color array
+        this.setState({ currentColor: color });
+        this.setState({ currentPosition: i });
+
+        this.setState({ displayPicker: !this.state.displayPicker });
+    }
+
+    /**
+     * Called to close the color picker from the color palette
+     */
+    onHide () {
+        this.setState({ displayPicker: !this.state.displayPicker });
+    }
+
+    /**
+     * Called when a user changes a color using the color picker. It affects the current color palette
+     *
+     * @param color - the new color
+     */
+    onChange (color) {
+        // Set the color picker to whatever new color the user has picked
+        let oldColor = this.state.currentColor;
+        let newColor = {
+            color: new Color(color.rgb),
+            count: 1
+        };
+        this.setState({ currentColor: newColor });
+
+        // Then update the current color array with the new color
+        let newColors = this.state.colors;
+        newColors[this.state.currentPosition] = newColor;
+        this.setState({ colors: newColors });
+
+        // Alert each individual widget to that a color has changed
+        // Each widget will have to check if the change applies to itself
+        EventEmitter.dispatch('color-palette:color-change', { old: oldColor.color, new: newColor.color });
+    }
+
+    /**
+     * Resets the color picker when a new scene has loaded
+     */
+    clearColors () {
+        this.setState({ colors: [] });
+    }
+
+    /**
+     * Official React lifecycle method
+     * Called every time state or props are changed
+     */
+    render () {
+        let colors = [];
+        if (this.state.colors) {
+            for (let i = 0; i < this.state.colors.length; i++) {
+                let color = this.state.colors[i];
+                let widgetStyle = { backgroundColor: color.color.getRgbaString() };
+
+                // This represents each squared div for a color in the color palette
+                colors.push(<div key={i} className='palette-color' onClick={ this.onClick.bind(null, color, i) }><div className='square' style={widgetStyle}></div></div>);
+            }
+        }
+
+        return (
+            <div>
+                <div className='colors'>
+                    { colors }
+                </div>
+                {/* Draggable modal */}
+                <Modal id='modal-test' dialogComponentClass={DraggableModal} x={this.state.x} y={this.state.y} enforceFocus={false} className='widget-modal' show={this.state.displayPicker} onHide={this.onHide}>
```

Comment:
I hadn't caught it previously, but a number of components are getting generated with an `id='modal-test'` attribute. Can we remove these? It doesn't look like anything is relying on these ids, and we don't want anyone accidentally depending on them in the future. Also, many instances of these components can co-exist at the same time and it's technically illegal to have multiple elements in the same DOM structure with the same ID.